[BUG] cloneset scale choose pod panic when diff > len(pods)

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!-->

**What happened**:
scale choose pod panic when diff > len(pods)

cloneset status: 
```bash
NAME                  DESIRED   UPDATED   UPDATED_READY   READY   TOTAL   AGE
default               4         0         0               5       5       2d2h
```
panic log:
```bash
I0702 10:23:05.325465       1 cloneset_scale.go:100] CloneSet default/test begin to scale in 1 pods including 0 (current rev)
I0702 10:23:05.325500       1 cloneset_controller.go:175] Finished syncing CloneSet default/test, cost 105.113125ms
E0702 10:23:05.325666       1 runtime.go:69] Observed a panic: runtime.boundsError{x:1, y:0, signed:true, code:0x2} (runtime error: slice bounds out of range [:1] with capacity 0)
/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:679
/usr/local/go/src/runtime/panic.go:93
/workspace/pkg/controller/cloneset/scale/cloneset_scale_util.go:104
/workspace/pkg/controller/cloneset/scale/cloneset_scale_util.go:114
/workspace/pkg/controller/cloneset/scale/cloneset_scale.go:103
/workspace/pkg/controller/cloneset/cloneset_controller.go:302
/workspace/pkg/controller/cloneset/cloneset_controller.go:254
/workspace/pkg/controller/cloneset/cloneset_controller.go:165
/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1357
panic: runtime error: slice bounds out of range [:1] with capacity 0 [recovered]
        panic: runtime error: slice bounds out of range [:1] with capacity 0
```
**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
1. create an cloneset with replicas=5 
2. update the cloneset's pod template and partition to 5
3. scale the clonset to replicas=4

**Anything else we need to know?**:

**Environment**:
- Kruise version: master
- Kubernetes version (use `kubectl version`): v1.17.3

--- a/pkg/controller/cloneset/scale/cloneset_scale_util.go
+++ b/pkg/controller/cloneset/scale/cloneset_scale_util.go
@@ -72,7 +72,7 @@ func calculateDiffs(cs *appsv1alpha1.CloneSet, revConsistent bool, totalPods int
 
 	if !revConsistent {
 		if cs.Spec.UpdateStrategy.Partition != nil {
-			currentRevDiff = notUpdatedPods - int(*cs.Spec.UpdateStrategy.Partition)
+			currentRevDiff = notUpdatedPods - integer.IntMin(int(*cs.Spec.UpdateStrategy.Partition), int(*cs.Spec.Replicas))
 		}
 
 		// Use maxSurge only if partition has not satisfied
@@ -94,12 +94,14 @@ func calculateDiffs(cs *appsv1alpha1.CloneSet, revConsistent bool, totalPods int
 func choosePodsToDelete(totalDiff int, currentRevDiff int, notUpdatedPods, updatedPods []*v1.Pod) []*v1.Pod {
 	choose := func(pods []*v1.Pod, diff int) []*v1.Pod {
 		// No need to sort pods if we are about to delete all of them.
-		// diff will always be <= len(filteredPods), so not need to handle > case.
 		if diff < len(pods) {
 			// Sort the pods in the order such that not-ready < ready, unscheduled
 			// < scheduled, and pending < running. This ensures that we delete pods
 			// in the earlier stages whenever possible.
 			sort.Sort(kubecontroller.ActivePods(pods))
+		} else if diff > len(pods) {
+			klog.Warningf("Diff > len(pods) in choosePodsToDelete func which is not expected.")
+			return pods
 		}
 		return pods[:diff]
 	}

--- a/pkg/controller/cloneset/scale/cloneset_scale_util_test.go
+++ b/pkg/controller/cloneset/scale/cloneset_scale_util_test.go
@@ -204,6 +204,39 @@ func TestCalculateDiffs(t *testing.T) {
 			expectedCurrentRevDiff: -2,
 		},
 		{
+			name: "scale in without revConsistent 6",
+			cs: &appsv1alpha1.CloneSet{
+				Spec: appsv1alpha1.CloneSetSpec{
+					Replicas: utilpointer.Int32Ptr(9),
+					UpdateStrategy: appsv1alpha1.CloneSetUpdateStrategy{
+						Partition: utilpointer.Int32Ptr(10),
+					},
+				},
+			},
+			revConsistent:          false,
+			totalPods:              10,
+			notUpdatedPods:         10,
+			expectedTotalDiff:      1,
+			expectedCurrentRevDiff: 1,
+		},
+		{
+			name: "scale in without revConsistent 7",
+			cs: &appsv1alpha1.CloneSet{
+				Spec: appsv1alpha1.CloneSetSpec{
+					Replicas: utilpointer.Int32Ptr(9),
+					UpdateStrategy: appsv1alpha1.CloneSetUpdateStrategy{
+						MaxSurge:  &intOrStr1,
+						Partition: utilpointer.Int32Ptr(10),
+					},
+				},
+			},
+			revConsistent:          false,
+			totalPods:              10,
+			notUpdatedPods:         10,
+			expectedTotalDiff:      0,
+			expectedCurrentRevDiff: 1,
+		},
+		{
 			name: "scale out with maxSurge < currentRevDiff",
 			cs: &appsv1alpha1.CloneSet{
 				Spec: appsv1alpha1.CloneSetSpec{

